File: general/0075-sort-colors.cpp
#include <utility>
#include <vector>
using namespace std;

class Solution {
public:
    // counting sort method
    void sortColors(vector<int>& nums) {
        vector<int> bookkeeper(3, 0);
        for (int num: nums) bookkeeper[num]++;

        int numsIter = 0;
        for (int i = 0; i < 3; i++) {
            while (bookkeeper[i] > 0) {
                nums[numsIter++] = i;
                bookkeeper[i]--;
            }
        }
    }

    // dutch national flag problem
    void sortColors_dnf(vector<int>& nums) {
        int lo = 0, mid = 0, hi = nums.size() - 1;
        while (mid <= hi) {
            if (nums[mid] == 0) {
                swap(nums[lo++], nums[mid++]);
            } else if (nums[mid] == 2) {
                swap(nums[hi--], nums[mid]);
            } else {
                mid++;
            }
        }
    }
};

File: neetcode/arrays-hashing/0075-sort-colors.py
class Solution:
    # 1. Simple and Standard Sorting Algorithms: O(nlogn)
    # 2. Counting Sort: TC O(n)
    # 3. Dutch National Flag
    def sortColors(self, nums: list[int]) -> None:
        freq = [0] * 3
        for num in nums:
            freq[num] += 1
        
        k = 0
        for i in range(3):
            while freq[i] > 0:
                nums[k] = i
                k += 1
                freq[i] -= 1

    def sortColorsDNF(self, nums: list[int]) -> None:
        lo = mid = 0
        hi = len(nums) - 1
        while mid <= hi:
            if nums[mid] == 0:
                nums[mid], nums[lo] = nums[lo], nums[mid]
                mid += 1
                lo += 1
            elif nums[mid] == 1:
                mid += 1
            else:
                nums[mid], nums[hi] = nums[hi], nums[mid]
                hi -= 1
